src/transformers/models/mra/modeling_mra.py

```diff
@@ -0,0 +1,1498 @@
+
+    return max_vals, max_vals_scatter
+
+
+def sparse_mask_B(mask, indices, block_size=32):
+    """
+    Converts attention mask to a sparse mask for high resolution logits.
+    """
+    assert len(mask.size()) == 2
+    assert len(indices.size()) == 2
+    assert mask.shape[0] == indices.shape[0]
+
+    batch_size, seq_len = mask.shape
+    num_block = seq_len // block_size
+
+    batch_idx = torch.arange(indices.size(0), dtype=torch.long, device=indices.device)
+    mask = mask.reshape(batch_size, num_block, block_size)
+    mask = mask[batch_idx[:, None], (indices % num_block).long(), :]
+
+    return mask
+
+
+def mm_to_sparse(dense_A, dense_B, indices, block_size=32):
+    """
+    Performs Sampled Dense Matrix Multiplication.
+    """
+    batch_size, A_size, dim = dense_A.size()
+    _, B_size, dim = dense_B.size()
+    assert A_size % block_size == 0
+    assert B_size % block_size == 0
+
+    dense_A = dense_A.reshape(batch_size, A_size // block_size, block_size, dim).transpose(-1, -2)
+    dense_B = dense_B.reshape(batch_size, B_size // block_size, block_size, dim).transpose(-1, -2)
+
+    assert len(dense_A.size()) == 4
+    assert len(dense_B.size()) == 4
+    assert len(indices.size()) == 2
+    assert dense_A.size(3) == 32
+    assert dense_B.size(3) == 32
+
+    dense_A = dense_A.contiguous()
+    dense_B = dense_B.contiguous()
+
+    indices = indices.int()
+    indices = indices.contiguous()
+
+    assert dense_A.is_contiguous()
+    assert dense_B.is_contiguous()
+    assert indices.is_contiguous()
+
+    return cuda_kernel.mm_to_sparse(dense_A, dense_B, indices.int())
+
+
+def sparse_dense_mm(sparse_A, indices, dense_B, A_num_block, block_size=32):
+    """
+    Performs matrix multiplication of a sparse matrix with a dense matrix.
+    """
+    batch_size, B_size, dim = dense_B.size()
+
+    assert B_size % block_size == 0
+    assert sparse_A.size(2) == block_size
+    assert sparse_A.size(3) == block_size
+
+    dense_B = dense_B.reshape(batch_size, B_size // block_size, block_size, dim).transpose(-1, -2)
+
+    assert len(sparse_A.size()) == 4
+    assert len(dense_B.size()) == 4
+    assert len(indices.size()) == 2
+    assert sparse_A.size(2) == 32
+    assert sparse_A.size(3) == 32
+    assert dense_B.size(3) == 32
+
+    sparse_A = sparse_A.contiguous()
+
+    indices = indices.int()
+    indices = indices.contiguous()
+    dense_B = dense_B.contiguous()
+
+    assert sparse_A.is_contiguous()
+    assert indices.is_contiguous()
+    assert dense_B.is_contiguous()
+
+    dense_C = cuda_kernel.sparse_dense_mm(sparse_A, indices, dense_B, A_num_block)
+    dense_C = dense_C.transpose(-1, -2).reshape(batch_size, A_num_block * block_size, dim)
+    return dense_C
+
+
+def transpose_indices(indices, dim_1_block, dim_2_block):
+    return ((indices % dim_2_block) * dim_1_block + torch.div(indices, dim_2_block, rounding_mode="floor")).long()
+
+
+class SampledDenseMM(torch.autograd.Function):
```

Comment:
Should be prefixed by Mra and maybe a better name for MM? What does it stand for.